### PR TITLE
A few rrules

### PR DESCRIPTION
**Pull request by willtebbutt**
A few basic rrules for non-dense matrices and related.

### CODE REVIEW

File: src/rules/linear_algebra/diagonal.jl [hunk 1]
```diff
@@ -0,0 +1,2 @@
+rrule(::typeof(Diagonal), d::AbstractVector) = Diagonal(d), Rule(ΔΩ->diag(ΔΩ))
```
Comment (ararslan) [inline src/rules/linear_algebra/diagonal.jl]:
```suggestion
rrule(::typeof(Diagonal), d::AbstractVector) = Diagonal(d), Rule(diag)
```

File: src/rules/linear_algebra/diagonal.jl [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+rrule(::typeof(diag), A::AbstractMatrix) = diag(A), Rule(ΔΩ->Diagonal(ΔΩ))
```
Comment (ararslan) [inline src/rules/linear_algebra/diagonal.jl]:
```suggestion
rrule(::typeof(diag), A::AbstractMatrix) = diag(A), Rule(Diagonal)
```

**Comment by willtebbutt** [comment]
> Doesn't matter much, but why not keep calling the directory linalg?

Figured it matched the stdlib name a bit better, but I also agree that it really doesn't matter.

**Comment by ararslan** [comment]
Personally I prefer linalg since it's concise and not particularly ambiguous, plus linear_algebra similarly doesn't fully coincide with the stdlib name because of the underscore.

**Comment by willtebbutt** [comment]
Have reverted to `linalg` -- I agree it's a better choice I think.

**Pull request merged**